internal/core/src/storage/OpenDALChunkManager.cpp

```diff
@@ -0,0 +1,204 @@
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <fstream>
+#include <string>
+
+#include "log/Log.h"
+#include "opendal.h"
+#include "common/EasyAssert.h"
+#include "storage/Util.h"
+#include "storage/OpenDALChunkManager.h"
+
+namespace milvus::storage {
+
+std::once_flag init_flag_;
+
+OpenDALChunkManager::OpenDALChunkManager(const StorageConfig& storage_config)
+    : default_bucket_name_(storage_config.bucket_name) {
+    std::call_once(init_flag_, []() { opendal_init_logger(); });
+    remote_root_path_ = storage_config.root_path;
+    std::string storageType;
+    if (storage_config.cloud_provider == "gcp") {
+        storageType = "gcs";
+    } else if (storage_config.cloud_provider == "aliyun") {
+        storageType = "oss";
+    } else if (storage_config.cloud_provider == "azure") {
+        storageType = "azblob";
+    } else {
+        storageType = "s3";
+    }
+
+    op_options_ = opendal_operator_options_new();
+    if (!storage_config.access_key_id.empty() &&
+        !storage_config.access_key_value.empty()) {
+        opendal_operator_options_set(
+            op_options_, "access_key_id", storage_config.access_key_id.c_str());
+        opendal_operator_options_set(op_options_,
+                                     "secret_access_key",
+                                     storage_config.access_key_value.c_str());
+        storageType = "s3";
+    }
+    opendal_operator_options_set(op_options_, "root", "/");
+    opendal_operator_options_set(
+        op_options_, "bucket", storage_config.bucket_name.c_str());
+    opendal_operator_options_set(op_options_,
+                                 "region",
+                                 storage_config.region.empty()
+                                     ? "us-east-1"
+                                     : storage_config.region.c_str());
+    opendal_operator_options_set(
+        op_options_,
+        "endpoint",
+        ((storage_config.useSSL ? "https://" : "http://") +
+         storage_config.address)
+            .c_str());
+    opendal_operator_options_set(
+        op_options_,
+        "enable_virtual_host_style",
+        storage_config.useVirtualHost ? "true" : "false");
+
+    op_ptr_ = opendal_operator_new(storageType.c_str(), op_options_);
+    AssertInfo(op_ptr_ != nullptr, "init opendal operator failed");
```

Comment:
The error handling seems to be weird.

How do we handle error here? through exception or what?  I don't see any detailed information about why the operator failed

Reply:
we need this issue https://github.com/apache/incubator-opendal/issues/3244